replayer/src/creplayer/creplayer.c

```diff
@@ -18,200 +22,188 @@
+  pid_t pid;
+
+  RETURN_NULL_ON_TRUE(!PyArg_ParseTuple(args, "i", &pid));
+
+  RAISE_EXCEPTION_WITH_ERRNO_ON_TRUE(ptrace(PTRACE_SYSCALL, pid, 0, 0) == -1);
+  waitpid(pid, 0, 0);
+
+  return Py_BuildValue("");
 }
 
-static PyObject* set_memory_in_replayed_process(PyObject *self, PyObject *args) {
+
+static PyObject* py_get_registers_from_tracee(PyObject *self, PyObject *args) {
+  pid_t pid;
+  struct user_regs_struct regs;
+
+  RETURN_NULL_ON_TRUE(!PyArg_ParseTuple(args, "i", &pid));
+  RAISE_EXCEPTION_WITH_ERRNO_ON_TRUE(ptrace(PTRACE_GETREGS, pid, 0, &regs) == -1);
+
+  return Py_BuildValue(
+      "(KKKKKKKKKKKKKKKKKKKKK)", regs.r15, regs.r14, regs.r13, regs.r12, regs.rbp, 
+      regs.rbx, regs.r11, regs.r10, regs.r9, regs.r8, regs.rax, regs.rcx, regs.rdx, regs.rsi,
+      regs.rdi, regs.orig_rax, regs.rip, regs.cs, regs.eflags, regs.rsp, regs.ss);
+}
+
+
+static PyObject* py_set_registers_in_tracee(PyObject *self, PyObject *args) {
+  pid_t pid;
+  struct user_regs_struct regs;
+
+  // TODO: do this less hacky... Right now we don't pass along all registers either way (to python and back), so we need to first get the registers from the tracee, 
+  // and then only set the relevant registers. 
+  // A probable solution is to always pass all registers back and forth 
+  RETURN_NULL_ON_TRUE(!PyArg_ParseTuple(args, "iKKKKKKKKKKKKKKKKKKKKK", &pid, &regs.r15, &regs.r14, &regs.r13, &regs.r12, &regs.rbp, &regs.rbx, &regs.r11, &regs.r10, &regs.r9, &regs.r8, 
+  &regs.rax, &regs.rcx, &regs.rdx, 
+  &regs.rsi, &regs.rdi, &regs.orig_rax, &regs.rip, &regs.cs, &regs.eflags, &regs.rsp, &regs.ss));
+  RAISE_EXCEPTION_WITH_ERRNO_ON_TRUE(ptrace(PTRACE_GETREGS, pid, 0, &regs) == -1);
+  RETURN_NULL_ON_TRUE(!PyArg_ParseTuple(args, "iKKKKKKKKKKKKKKKKKKKKK", &pid, &regs.r15, &regs.r14, &regs.r13, &regs.r12, &regs.rbp, &regs.rbx, &regs.r11, &regs.r10, 
+  &regs.r9, &regs.r8, &regs.rax, &regs.rcx, &regs.rdx, &regs.rsi, &regs.rdi, &regs.orig_rax, &regs.rip, &regs.cs, &regs.eflags, &regs.rsp, &regs.ss));
+  RAISE_EXCEPTION_WITH_ERRNO_ON_TRUE(ptrace(PTRACE_SETREGS, pid, 0, &regs) == -1);
+
+  return Py_BuildValue("");
+}
+
+
+static PyObject* py_write_to_tracee(PyObject *self, PyObject *args) {
+  pid_t pid;
   unsigned long long address;
   char *buffer;
   long length;
 
-  RETURN_NULL_ON_TRUE(!PyArg_ParseTuple(args, "Ks#", &address, &buffer, &length));
-  RETURN_NULL_ON_TRUE(write_buffer_to_tracee(address, buffer, length) == NULL);
+  RETURN_NULL_ON_TRUE(!PyArg_ParseTuple(args, "iKs#", &pid, &address, &buffer, &length));
+  RETURN_NULL_ON_TRUE(write_buffer_to_tracee(pid, address, buffer, length) == NULL);
 
   return Py_BuildValue("");
-} 
+}
+
 
+static PyObject* py_read_from_tracee(PyObject *self, PyObject *args) {
+  pid_t pid;
+  unsigned long long address;
+  int length;
+  char *buffer;
+
+  RETURN_NULL_ON_TRUE(!PyArg_ParseTuple(args, "iKi", &pid, &address, &length));
+  buffer = malloc(length);
```

Comment:
Why would malloc ever fail

Reply:
WHO FREES THIS??!

Reply:
I'm pretty sure I'm not supposed to free this, since I pass it along to python- the garbage collector will free once it's reference count falls to 0, no?